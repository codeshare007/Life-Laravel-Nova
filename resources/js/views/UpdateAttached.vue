<template>
    <loading-view :loading="loading">
        <heading class="mb-3">{{__('Update')}} {{ relatedResourceLabel }}</heading>

        <card class="overflow-hidden">
            <form v-if="field" @submit.prevent="updateAttachedResource" autocomplete="off">
                <!-- Related Resource -->
                <default-field :field="field" :errors="validationErrors">
                    <template slot="field">
                       <select
                            class="form-control form-select mb-3 w-full"
                            dusk="attachable-select"
                            :class="{ 'border-danger': validationErrors.has(field.attribute) }"
                            :data-testid="`${field.resourceName}-select`"
                            @change="selectResourceFromSelectControl"
                            disabled
                        >
                            <option value="" disabled selected>{{__('Choose')}} {{ field.name }}</option>

                            <option
                                v-for="resource in availableResources"
                                :key="resource.value"
                                :value="resource.value"
                                :selected="selectedResourceId == resource.value"
                            >
                                {{ resource.display}}
                            </option>
                        </select>
                    </template>
                </default-field>

                <!-- Pivot Fields -->
                <div v-for="field in fields">
                    <component
                        :is="'form-' + field.component"
                        :resource-name="resourceName"
                        :resource-id="resourceId"
                        :field="field"
                        :errors="validationErrors"
                        :related-resource-name="relatedResourceName"
                        :related-resource-id="relatedResourceId"
                        :via-resource="viaResource"
                        :via-resource-id="viaResourceId"
                        :via-relationship="viaRelationship"
                    />
                </div>

                <!-- Attach Button -->
                <div class="bg-30 flex px-8 py-4">
                    <button dusk="update-and-continue-editing-button" type="button" @click="updateAndContinueEditing" class="ml-auto btn btn-default btn-primary mr-3">
                        {{__('Update & Continue Editing')}}
                    </button>

                    <button dusk="update-button" class="btn btn-default btn-primary">
                        {{__('Update')}} {{ relatedResourceLabel }}
                    </button>
                </div>
            </form>
        </card>
    </loading-view>
</template>

<script>
import _ from 'lodash'
import { PerformsSearches, TogglesTrashed, Errors } from 'laravel-nova'

export default {
    mixins: [PerformsSearches, TogglesTrashed],

    props: {
        resourceName: {
            type: String,
            required: true,
        },
        resourceId: {
            required: true,
        },
        relatedResourceName: {
            type: String,
            required: true,
        },
        relatedResourceId: {
            required: true,
        },
        viaResource: {
            default: '',
        },
        viaResourceId: {
            default: '',
        },
        viaRelationship: {
            default: '',
        },
        polymorphic: {
            default: false,
        },
    },

    data: () => ({
        loading: true,
        field: null,
        softDeletes: false,
        fields: [],
        validationErrors: new Errors(),
        selectedResource: null,
        selectedResourceId: null,
        lastRetrievedAt: null,
    }),

    /**
     * Mount the component.
     */
    mounted() {
        this.initializeComponent()
    },

    methods: {
        /**
         * Initialize the component's data.
         */
        async initializeComponent() {
            this.softDeletes = false
            this.disableWithTrashed()
            this.clearSelection()
            this.getField()

            await this.getPivotFields()
            await this.getAvailableResources()

            this.selectedResourceId = this.relatedResourceId

            this.selectInitialResource()

            this.updateLastRetrievedAtTimestamp()
        },

        /**
         * Get the many-to-many relationship field.
         */
        getField() {
            this.field = null

            Nova.request()
                .get('/nova-api/' + this.resourceName + '/field/' + this.viaRelationship)
                .then(({ data }) => {
                    this.field = data

                    if (this.field.searchable) {
                        this.determineIfSoftDeletes()
                    }

                    this.loading = false
                })
        },

        /**
         * Get all of the available pivot fields for the relationship.
         */
        async getPivotFields() {
            this.fields = []

            const { data } = await Nova.request()
                .get(
                    `/nova-api/${this.resourceName}/${this.resourceId}/update-pivot-fields/${
                        this.relatedResourceName
                    }/${this.relatedResourceId}`,
                    { params: { viaRelationship: this.viaRelationship } }
                )
                .catch(error => {
                    if (error.response.status == 404) {
                        this.$router.push({ name: '404' })
                        return
                    }
                })

            this.fields = data

            _.each(this.fields, field => {
                field.fill = () => ''
            })
        },

        /**
         * Get all of the available resources for the current search / trashed state.
         */
        async getAvailableResources(search = '') {
            try {
                const response = await Nova.request().get(
                    `/nova-api/${this.resourceName}/${this.resourceId}/attachable/${
                        this.relatedResourceName
                    }`,
                    {
                        params: {
                            search,
                            current: this.relatedResourceId,
                            first: true,
                            withTrashed: this.withTrashed,
                        },
                    }
                )

                this.availableResources = response.data.resources
                this.withTrashed = response.data.withTrashed
                this.softDeletes = response.data.softDeletes
            } catch (error) {
                console.log(error)
            }
        },

        /**
         * Determine if the related resource is soft deleting.
         */
        determineIfSoftDeletes() {
            Nova.request()
                .get('/nova-api/' + this.relatedResourceName + '/soft-deletes')
                .then(response => {
                    this.softDeletes = response.data.softDeletes
                })
        },

        /**
         * Update the attached resource.
         */
        async updateAttachedResource() {
            try {
                await this.updateRequest()

                this.$toasted.show(this.__('The resource was updated!'), { type: 'success' })

                this.$router.push({
                    name: 'detail',
                    params: {
                        resourceName: this.resourceName,
                        resourceId: this.resourceId,
                    },
                })
            } catch (error) {
                console.log(error)
                if (error.response.status == 422) {
                    this.validationErrors = new Errors(error.response.data.errors)
                }

                if (error.response.status == 409) {
                    this.$toasted.show(
                        this.__(
                            'Another user has updated this resource since this page was loaded. Please refresh the page and try again.'
                        ),
                        { type: 'error' }
                    )
                }
            }
        },

        /**
         * Update the resource and reset the form
         */
        async updateAndContinueEditing() {
            try {
                await this.updateRequest()

                this.$toasted.show(this.__('The resource was updated!'), { type: 'success' })

                // Reset the form by refetching the fields
                this.initializeComponent()
            } catch (error) {
                console.log(error)
                if (error.response.status == 422) {
                    this.validationErrors = new Errors(error.response.data.errors)
                }

                if (error.response.status == 409) {
                    this.$toasted.show(
                        this.__(
                            'Another user has updated this resource since this page was loaded. Please refresh the page and try again.'
                        ),
                        { type: 'error' }
                    )
                }
            }
        },

        /**
         * Send an update request for this resource
         */
        updateRequest() {
            return Nova.request().post(
                `/nova-api/${this.resourceName}/${this.resourceId}/update-attached/${
                    this.relatedResourceName
                }/${this.relatedResourceId}`,
                this.updateAttachmentFormData
            )
        },

        /**
         * Select a resource using the <select> control
         */
        selectResourceFromSelectControl(e) {
            this.selectedResourceId = e.target.value
            console.log(e.target.value, this.selectedResourceId)
            this.selectInitialResource()
        },

        /**
         * Toggle the trashed state of the search
         */
        toggleWithTrashed() {
            this.withTrashed = !this.withTrashed

            // Reload the data if the component doesn't support searching
            if (!this.isSearchable) {
                this.getAvailableResources()
            }
        },

        /**
         * Select the initial selected resource
         */
        selectInitialResource() {
            this.selectedResource = _.find(
                this.availableResources,
                r => r.value == this.selectedResourceId
            )
        },

        /**
         * Update the last retrieved at timestamp to the current UNIX timestamp.
         */
        updateLastRetrievedAtTimestamp() {
            this.lastRetrievedAt = Math.floor(new Date().getTime() / 1000)
        },
    },

    computed: {
        /**
         * Get the attachment endpoint for the relationship type.
         */
        attachmentEndpoint() {
            return this.polymorphic
                ? '/nova-api/' +
                      this.resourceName +
                      '/' +
                      this.resourceId +
                      '/attach-morphed/' +
                      this.relatedResourceName
                : '/nova-api/' +
                      this.resourceName +
                      '/' +
                      this.resourceId +
                      '/attach/' +
                      this.relatedResourceName
        },

        /*
         * Get the form data for the resource attachment update.
         */
        updateAttachmentFormData() {
            return _.tap(new FormData(), formData => {
                _.each(this.fields, field => {
                    field.fill(formData)
                })

                formData.append('viaRelationship', this.viaRelationship)

                if (!this.selectedResource) {
                    formData.append(this.relatedResourceName, '')
                } else {
                    formData.append(this.relatedResourceName, this.selectedResource.value)
                }

                formData.append(this.relatedResourceName + '_trashed', this.withTrashed)
                formData.append('_retrieved_at', this.lastRetrievedAt)
            })
        },

        /**
         * Get the label for the related resource.
         */
        relatedResourceLabel() {
            if (this.field) {
                return this.field.singularLabel
            }
        },

        /**
         * Determine if the related resources is searchable
         */
        isSearchable() {
            return this.field.searchable
        },
    },
}
</script>
