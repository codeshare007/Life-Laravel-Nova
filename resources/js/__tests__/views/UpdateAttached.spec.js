import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import UpdateAttached from '@/views/UpdateAttached'
// import flushPromises from 'flush-promises'

describe('UpdateAttached', () => {
    test('it loads all the available resources if its not searchable', () => {
        window.Nova = {}
        window.Nova.config = {
            resources: [
                {
                    uriKey: 'users',
                    label: 'Users',
                    singularLabel: 'User',
                    authorizedToCreate: true,
                    searchable: false,
                },
                {
                    uriKey: 'roles',
                    label: 'Roles',
                    singularLabel: 'Role',
                    authorizedToCreate: true,
                    searchable: false,
                },
            ],
        }

        const wrapper = mount(UpdateAttached, {
            propsData: {
                resourceName: 'users',
                resourceId: 100,
                relatedResourceName: 'roles',
                relatedResourceId: 25,
                // viaResource: {},
                // viaResourceId: {},
                // viaRelationship: {},
                // polymorphic: false,
            },
        })

        wrapper.setData({
            field: {},
        })

        // expect(wrapper).toMatchSnapshot()
    })
})
