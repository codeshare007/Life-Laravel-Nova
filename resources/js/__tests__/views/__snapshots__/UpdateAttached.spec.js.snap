// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAttached it loads all the available resources if its not searchable 1`] = `<div class="card overflow-hidden"></div>`;
